File: javascript/ch1/Ex122.js
// Load with: const ExXXX = require('./ExXXX')
// Run with: ExXXX.xxx()

let sq = n => n * n;

let divides = (a, b) => (b % a) == 0;

let find_divisor = (n, test_divisor) => {
    if (sq(test_divisor) > n) {
        return n;
    } else if (divides(test_divisor, n)) {
        return test_divisor;
    } else {
        return find_divisor(n, test_divisor + 1);
    }
};

let smallest_divisor = n => find_divisor(n, 2);

let odd = n => !divides(2, n);

let is_prime = n => n == smallest_divisor(n);

let prime_test = n => {
    if (is_prime(n)) {
        console.log(n + " ***");
    } else {
        console.log(n);
    }
};

let search_helper = (n, end) => {
    if (n <= end) {
        prime_test(n);
        search_helper(n + 2, end);
    } else {
        console.log("finished\n");
    }
};

let search_for_primes = (start, end) => search_helper(odd(start) ? start : start + 1, end);

exports.search_for_primes = search_for_primes;

exports.test_it = function () {
    search_for_primes(1000, 1020);
    search_for_primes(10000, 10040);
    search_for_primes(100000, 100080);
};

File: haskell/ch1/Ex122.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...

module Ex122 where

sq n = n * n

divides a b = (mod b a) == 0

find_divisor n test | sq test > n = n
find_divisor n test | divides test n = test
find_divisor n test = find_divisor n (test + 1)

smallest_divisor n = find_divisor n 2

is_prime n = (smallest_divisor n) == n

prime_test n | is_prime n = show n ++ " ***"
prime_test n = show n

range start end = [start, (start + 2) .. end]

make_range start end = if odd start
                       then range start end
                       else range (start + 1) end

report vals = mapM_ putStrLn vals

search_for_primes start end = report (map prime_test (make_range start end))
